Documentation updates, mainly for Python API (#5531)
apache/pulsar: site2/docs/functions-develop.md

```
For complete code, see [here](https://github.com/apache/pulsar/blob/master/pulsar-functions/python-examples/native_exclamation_function.py).

Note that functions can be written in python2 or python3, but pulsar
currently only looks for "python" as the interpreter to execute them.

A recent Ubuntu system may have only "python3" but not "python", in which
case functions will fail to start.  As a workaround you can create a symlink, but beware this has some
[risks](https://askubuntu.com/questions/320996/how-to-make-python-program-command-execute-python-3#answer-475815):

```bash
sudo update-alternatives --install /usr/bin/python python /usr/bin/python3 10
```

If you choose to do this, be careful not to install any other package which

> depends on "python" (2.x)

Reviewer: > Note 
> You can write Pulsar Functions in python2 or python3. However, Pulsar only looks for `python` as the interpreter.

> If you're running Pulsar Functions on Ubuntu system that only supports `python3`, you might fail to
start the functions. In this case, you can create a symlink. However, such action
has potential [risks](https://askubuntu.com/questions/320996/how-to-make-python-program-command-execute-python-3#answer-475815).

```bash
sudo update-alternatives --install /usr/bin/python python /usr/bin/python3 10
```
> If you create a symlink, you'd better not install any other package that depends on "python" (2.x).

Reviewer: A tiny suggestion on this: If we can summarize the risks briefly, we can use a brief summary, and not use external links.
Reason: The external link might change or someone might delete the Q&A in this link, then the link in our docs will be broken, it's hard to maintain external links.

Author: It could link to #5518 on github instead.  Actually, I'm hoping #5518 will accept a new config option for setting the path to the python interpreter, in which case we would document that instead.

Reviewer: yes, sure.

Pulsar Functions use [Apache BookKeeper](https://bookkeeper.apache.org) as a state storage interface. Pulsar installation, including the local standalone installation, includes deployment of BookKeeper bookies.

Since Pulsar 2.1.0 release, Pulsar integrates with Apache BookKeeper [table service](https://docs.google.com/document/d/155xAwWv5IdOitHh1NVMEwCMGgB28M3FyMiQSxEpjE-Y/edit#heading=h.56rbh52koe3f) to store the `State` for functions. For example, a `WordCount` function can store its `counters` state into BookKeeper table service via Pulsar Functions State API.

> States are key-value pairs, where the key is a string and the value is arbitrary binary data - counters are stored as 64-bit big-endian binary values.  Keys are scoped to an individual pulsar function, but shared between all instances of that function.

Reviewer: ```suggestion
States are key-value pairs, where the key is a string and the value is arbitrary binary data - counters are stored as 64-bit big-endian binary values. Thought keys are scoped to an individual Pulsar Function, the keys are shared among all instances of that function.
```

Author: s/Thought/Though/ ?

I think the conjunction isn't useful here, as it implies the user already knows the scoping of keys.  I don't think it's mentioned earlier, and without this knowledge I might have guessed they were scoped differently (to the pulsar "namespace" that the function executes within, for example)

So as a user, I just want a statement which answers the question: "What's the scope of the key?"

I would be happy with "Keys are scoped to the pulsar function".  I thought it worth clarifying functions versus function instances, but maybe that's unnecessary.  State storage which *wasn't* shared between function instances wouldn't be very useful.

Reviewer: "Though", sorry for the typo.
If you think the conjunction is useful, you can remove it.

> States are accessed within pulsar functions using the `putState`, `getState`, `incrCounter`, `getCounter` and `deleteState` calls on the context object.  They can also be managed using the [querystate](pulsar-admin.md#querystate) and [putstate](pulsar-admin.md#putstate) options to `pulsar-admin functions`.

Reviewer: ```suggestion
You can access states within Pulsar Functions using the `putState`, `getState`, `incrCounter`, `getCounter` and `deleteState` calls on the context object. You can also manage states using the [querystate](pulsar-admin.md#querystate) and [putstate](pulsar-admin.md#putstate) options to `pulsar-admin functions`.
```

Author: I'll accept that - active rather than passive.
